sdks/java/io/snowflake/src/main/java/org/apache/beam/sdk/io/snowflake/SnowflakeIO.java

```diff
@@ -447,6 +513,346 @@ public void populateDisplayData(DisplayData.Builder builder) {
+                        @Override
+                        public Object[] apply(T element) {
+                          return getUserDataMapper().mapRow(element);
+                        }
+                      }))
+              .apply("Map Objects array to CSV lines", ParDo.of(new MapObjectsArrayToCsvFn()))
+              .setCoder(StringUtf8Coder.of());
+
+      WriteFilesResult filesResult =
+          mappedUserData.apply(
+              "Write files to specified location",
+              FileIO.<String>write()
+                  .via(TextIO.sink())
+                  .to(stagingBucketDir)
+                  .withPrefix(getFileNameTemplate())
+                  .withSuffix(".csv")
+                  .withCompression(Compression.GZIP));
+
+      return (PCollection)
+          filesResult
+              .getPerDestinationOutputFilenames()
+              .apply("Parse KV filenames to Strings", Values.<String>create());
+    }
+
+    private ParDo.SingleOutput<Object, Object> copyToTable(
+        SnowflakeService snowflakeService, String stagingBucketDir) {
+      return ParDo.of(
+          new CopyToTableFn<>(
+              getDataSourceProviderFn(),
+              getTable(),
+              getQuery(),
+              stagingBucketDir,
+              getStorageIntegrationName(),
+              getWriteDisposition(),
+              snowflakeService));
+    }
+  }
+
+  public static class Concatenate extends Combine.CombineFn<String, List<String>, List<String>> {
+    @Override
+    public List<String> createAccumulator() {
+      return new ArrayList<>();
+    }
+
+    @Override
+    public List<String> addInput(List<String> mutableAccumulator, String input) {
+      mutableAccumulator.add(String.format("'%s'", input));
+      return mutableAccumulator;
+    }
+
+    @Override
+    public List<String> mergeAccumulators(Iterable<List<String>> accumulators) {
+      List<String> result = createAccumulator();
+      for (List<String> accumulator : accumulators) {
+        result.addAll(accumulator);
+      }
+      return result;
+    }
+
+    @Override
+    public List<String> extractOutput(List<String> accumulator) {
+      return accumulator;
+    }
+  }
+
+  /**
+   * Custom DoFn that maps {@link Object[]} into CSV line to be saved to Snowflake.
+   *
+   * <p>Adds Snowflake-specific quotations around strings.
+   */
+  private static class MapObjectsArrayToCsvFn extends DoFn<Object[], String> {
+
+    @ProcessElement
+    public void processElement(ProcessContext context) {
+      List<Object> csvItems = new ArrayList<>();
+      for (Object o : context.element()) {
+        if (o instanceof String) {
+          String field = (String) o;
+          field = field.replace("'", "''");
+          field = quoteField(field);
+
+          csvItems.add(field);
+        } else {
+          csvItems.add(o);
+        }
+      }
+      context.output(Joiner.on(",").useForNull("").join(csvItems));
```

Comment:
This is where there's an implicit `.toString()` on the objects in the array -- this is still pretty dangerous for many non-primitive classes!

I can think of a couple of solutions: (1) Use something other than CSV as the file for copying into Snowflake, (2) Add a warning to the data mapper doc that the toString's have to be coherent!

For example, if my data mapper function returns `new Object[] {Arrays.asList(1,',',"\n")}` to insert, it's almost certainly going to break the function.

I can see this happening, for example, if a user thinks that returning a JsonObject will insert the JSON as a string into that column.

Reply:
@RyanSkraba thanks for pointing this. We're going to add this concern to our feature improvement list :)

Reply:
Thanks for pointing this. We're going to add it to our feature improvement list.